File: src/main/java/com/manson/fo76/service/UserService.java
package com.manson.fo76.service;

import com.manson.fo76.domain.dto.User;
import com.manson.fo76.repository.UserRepository;
import java.util.ArrayList;
import java.util.Optional;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.stereotype.Service;

@Slf4j
@Service
public class UserService implements UserDetailsService {

  private final UserRepository userRepository;

  @Autowired
  public UserService(UserRepository userRepository) {
    this.userRepository = userRepository;
  }

  public User findById(String id) {
    if (StringUtils.isNotBlank(id)) {
      return userRepository.findById(id).orElse(null);
    }
    return null;
  }

  public User findByName(String name) {
    if (StringUtils.isNotBlank(name)) {
      return userRepository.findByName(name);
    }
    return null;
  }

  public User findByIdOrName(String idOrName) {
    return Optional.ofNullable(findById(idOrName)).orElse(findByName(idOrName));
  }

  @Override
  public UserDetails loadUserByUsername(String name) {
    User user = findByName(name);
    if (user != null) {
      return new org.springframework.security.core.userdetails.User(
          user.getName(), user.getPassword(),
          new ArrayList<>()
      );
    }
    return null;
  }
}

File: src/main/java/com/manson/fo76/service/UserService.kt
package com.manson.fo76.service

import com.manson.fo76.domain.dto.User
import com.manson.fo76.helper.Utils
import com.manson.fo76.repository.UserRepository
import java.util.Objects
import java.util.Optional
import java.util.UUID
import org.apache.commons.lang3.StringUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service

@Service
class UserService @Autowired constructor(private val userRepository: UserRepository) : UserDetailsService {
    fun createNewUser(user: User): User? {
        val userInDb = userRepository.findByName(user.name)
        if (Objects.isNull(userInDb)) {
            user.password = UUID.randomUUID().toString()
            return userRepository.save(user)
        }
        return null
    }

    fun findById(id: String): User? {
        return if (StringUtils.isNotBlank(id)) {
            userRepository.findById(id).orElse(null)
        } else null
    }

    fun findByName(name: String?): User? {
        return if (StringUtils.isNotBlank(name)) {
            userRepository.findByName(name)
        } else null
    }

    fun findByIdOrName(idOrName: String?): User {
        return Optional.ofNullable(findById(idOrName!!)).orElseGet { userRepository.findByName(idOrName) }
    }

    fun findByIdOrName(user: User): User? {
        return if (Objects.nonNull(user)) {
            Optional.ofNullable(findByIdOrName(user.id))
                    .orElseGet { findByIdOrName(user.name) }
        } else null
    }

    fun findAll(): List<User?> {
        return userRepository.findAll()
    }

    fun deleteAll() {
        userRepository.deleteAll()
    }

    fun delete(user: User): User? {
        if (Objects.isNull(user)) {
            return null
        }
        val userInDb = findByIdOrName(user)
        if (Objects.isNull(userInDb)) {
            return null
        }
        if (StringUtils.equals(userInDb!!.password, user.password)) {
            userRepository.delete(userInDb)
            return userInDb
        }
        return null
    }

    override fun loadUserByUsername(name: String): UserDetails? {
        val user = findByName(name)
        return if (user != null) {
            org.springframework.security.core.userdetails.User(user.name, user.password,
                    listOf())
        } else null
    }

    fun loadUser(user: User): UserDetails? {
        if (Objects.isNull(user)) {
            return null
        }
        val userInDb = findByIdOrName(user)
        return if (userInDb != null && Utils.validatePassword(user, userInDb)) {
            org.springframework.security.core.userdetails.User(userInDb.name, userInDb.password,
                    listOf())
        } else null
    }
}